### PR TITLE
sd-netlink: add sizes for IP and Ethernet addresses and sockets

### PR DESCRIPTION
**Pull request by topimiettinen**
Add sizes for uses of NETLINK_TYPE_IN_ADDR, NETLINK_TYPE_SOCKADDR and
NETLINK_TYPE_ETHER_ADDR types.

### CODE REVIEW

**Comment by topimiettinen** [comment]
Not sure what's the significance of `.size` attribute, it seems to be mostly ignored. Also in some cases the protocols can use IPv4 and IPv6 addresses using the same attribute.

**Comment by topimiettinen** [comment]
If the attribute is indeed ignored, perhaps the existing `.size` definitions should be removed instead.

**Comment by topimiettinen** [comment]
rpm-i386 fails early with `ImportError: libffi.so.8: cannot open shared object file: No such file or directory`, not related

File: src/libsystemd/sd-netlink/netlink-types-genl.c [hunk 1]
```diff
@@ -194,7 +194,7 @@ static const NLType genl_nl80211_types[] = {
 /***************** genl wireguard type systems *****************/
 static const NLType genl_wireguard_allowedip_types[] = {
         [WGALLOWEDIP_A_FAMILY]    = { .type = NETLINK_TYPE_U16 },
-        [WGALLOWEDIP_A_IPADDR]    = { .type = NETLINK_TYPE_IN_ADDR },
+        [WGALLOWEDIP_A_IPADDR]    = { .type = NETLINK_TYPE_IN_ADDR, .size = sizeof(struct in_addr) },
```
Comment (yuwata) [inline src/libsystemd/sd-netlink/netlink-types-genl.c]: This takes both IPv4 and IPv6.

File: src/libsystemd/sd-netlink/netlink-types-genl.c [hunk 2]
```diff
@@ -205,7 +205,7 @@ static const NLType genl_wireguard_peer_types[] = {
         [WGPEER_A_FLAGS]                         = { .type = NETLINK_TYPE_U32 },
         [WGPEER_A_PRESHARED_KEY]                 = { .type = NETLINK_TYPE_BINARY, .size = WG_KEY_LEN },
         [WGPEER_A_PERSISTENT_KEEPALIVE_INTERVAL] = { .type = NETLINK_TYPE_U16 },
-        [WGPEER_A_ENDPOINT]                      = { .type = NETLINK_TYPE_SOCKADDR },
+        [WGPEER_A_ENDPOINT]                      = { .type = NETLINK_TYPE_SOCKADDR, .size = sizeof(struct sockaddr_in) },
```
Comment (yuwata) [inline src/libsystemd/sd-netlink/netlink-types-genl.c]: This takes sockaddr_in or sockaddr_in6. Please drop the change.

File: src/libsystemd/sd-netlink/netlink-types-rtnl.c [hunk 3]
```diff
@@ -216,8 +216,8 @@ static const NLType rtnl_link_info_data_gre_types[] = {
         [IFLA_GRE_OFLAGS]           = { .type = NETLINK_TYPE_U16 },
         [IFLA_GRE_IKEY]             = { .type = NETLINK_TYPE_U32 },
         [IFLA_GRE_OKEY]             = { .type = NETLINK_TYPE_U32 },
-        [IFLA_GRE_LOCAL]            = { .type = NETLINK_TYPE_IN_ADDR },
-        [IFLA_GRE_REMOTE]           = { .type = NETLINK_TYPE_IN_ADDR },
+        [IFLA_GRE_LOCAL]            = { .type = NETLINK_TYPE_IN_ADDR, .size = sizeof(struct in_addr) },
+        [IFLA_GRE_REMOTE]           = { .type = NETLINK_TYPE_IN_ADDR, .size = sizeof(struct in_addr) },
```
Comment (yuwata) [inline src/libsystemd/sd-netlink/netlink-types-rtnl.c]: These take both IPv4 and IPv6. Please drop the change.

File: src/libsystemd/sd-netlink/netlink-types-rtnl.c [hunk 4]
```diff
@@ -248,8 +248,10 @@ static const NLType rtnl_link_info_data_ipoib_types[] = {
 #define DEFINE_IPTUN_TYPES(name, flags_type)                                            \
         static const NLType rtnl_link_info_data_##name##_types[] = {                    \
                 [IFLA_IPTUN_LINK]                = { .type = NETLINK_TYPE_U32 },        \
-                [IFLA_IPTUN_LOCAL]               = { .type = NETLINK_TYPE_IN_ADDR },    \
-                [IFLA_IPTUN_REMOTE]              = { .type = NETLINK_TYPE_IN_ADDR },    \
+                [IFLA_IPTUN_LOCAL]               = { .type = NETLINK_TYPE_IN_ADDR,      \
+                                                     .size = sizeof(struct in_addr) },  \
+                [IFLA_IPTUN_REMOTE]              = { .type = NETLINK_TYPE_IN_ADDR,      \
+                                                     .size = sizeof(struct in_addr) },  \
```
Comment (yuwata) [inline src/libsystemd/sd-netlink/netlink-types-rtnl.c]: These also take IPv4 and IPv6.

File: src/libsystemd/sd-netlink/netlink-types-rtnl.c [hunk 5]
```diff
@@ -349,8 +351,8 @@ static const NLType rtnl_link_info_data_vti_types[] = {
         [IFLA_VTI_LINK]         = { .type = NETLINK_TYPE_U32 },
         [IFLA_VTI_IKEY]         = { .type = NETLINK_TYPE_U32 },
         [IFLA_VTI_OKEY]         = { .type = NETLINK_TYPE_U32 },
-        [IFLA_VTI_LOCAL]        = { .type = NETLINK_TYPE_IN_ADDR },
-        [IFLA_VTI_REMOTE]       = { .type = NETLINK_TYPE_IN_ADDR },
+        [IFLA_VTI_LOCAL]        = { .type = NETLINK_TYPE_IN_ADDR, .size = sizeof(struct in_addr) },
+        [IFLA_VTI_REMOTE]       = { .type = NETLINK_TYPE_IN_ADDR, .size = sizeof(struct in_addr) },
```
Comment (yuwata) [inline src/libsystemd/sd-netlink/netlink-types-rtnl.c]: Same here. Please drop.

File: src/libsystemd/sd-netlink/netlink-types-rtnl.c [hunk 6]
```diff
@@ -839,8 +841,8 @@ static const NLType rtnl_proto_down_reason_types[] = {
 DEFINE_TYPE_SYSTEM(rtnl_proto_down_reason);
 
 static const NLType rtnl_link_types[] = {
-        [IFLA_ADDRESS]             = { .type = NETLINK_TYPE_ETHER_ADDR },
-        [IFLA_BROADCAST]           = { .type = NETLINK_TYPE_ETHER_ADDR },
+        [IFLA_ADDRESS]             = { .type = NETLINK_TYPE_ETHER_ADDR, .size = ETH_ALEN },
+        [IFLA_BROADCAST]           = { .type = NETLINK_TYPE_ETHER_ADDR, .size = ETH_ALEN },
```
Comment (yuwata) [inline src/libsystemd/sd-netlink/netlink-types-rtnl.c]: These depends on the hardware, e.g. INFINIBAND_ALEN for infiniband interfaces. Please drop the change.

File: src/libsystemd/sd-netlink/netlink-types-rtnl.c [hunk 7]
```diff
@@ -891,7 +893,7 @@ static const NLType rtnl_link_types[] = {
         [IFLA_MAX_MTU]             = { .type = NETLINK_TYPE_U32 },
         [IFLA_PROP_LIST]           = { .type = NETLINK_TYPE_NESTED, .type_system = &rtnl_prop_list_type_system },
         [IFLA_ALT_IFNAME]          = { .type = NETLINK_TYPE_STRING, .size = ALTIFNAMSIZ - 1 },
-        [IFLA_PERM_ADDRESS]        = { .type = NETLINK_TYPE_ETHER_ADDR },
+        [IFLA_PERM_ADDRESS]        = { .type = NETLINK_TYPE_ETHER_ADDR, .size = ETH_ALEN },
```
Comment (yuwata) [inline src/libsystemd/sd-netlink/netlink-types-rtnl.c]: Also here. Please drop.

File: src/libsystemd/sd-netlink/netlink-types-rtnl.c [hunk 8]
```diff
@@ -902,13 +904,13 @@ DEFINE_TYPE_SYSTEM(rtnl_link);
 /* IFA_FLAGS was defined in kernel 3.14, but we still support older
  * kernels where IFA_MAX is lower. */
 static const NLType rtnl_address_types[] = {
-        [IFA_ADDRESS]           = { .type = NETLINK_TYPE_IN_ADDR },
-        [IFA_LOCAL]             = { .type = NETLINK_TYPE_IN_ADDR },
+        [IFA_ADDRESS]           = { .type = NETLINK_TYPE_IN_ADDR, .size = sizeof(struct in_addr) },
+        [IFA_LOCAL]             = { .type = NETLINK_TYPE_IN_ADDR, .size = sizeof(struct in_addr) },
```
Comment (yuwata) [inline src/libsystemd/sd-netlink/netlink-types-rtnl.c]: These take both IPv4 and IPv6.

File: src/libsystemd/sd-netlink/netlink-types-rtnl.c [hunk 9]
```diff
@@ -902,13 +904,13 @@ DEFINE_TYPE_SYSTEM(rtnl_link);
...
         [IFA_LABEL]             = { .type = NETLINK_TYPE_STRING, .size = IFNAMSIZ - 1 },
-        [IFA_BROADCAST]         = { .type = NETLINK_TYPE_IN_ADDR },
-        [IFA_ANYCAST]           = { .type = NETLINK_TYPE_IN_ADDR },
+        [IFA_BROADCAST]         = { .type = NETLINK_TYPE_IN_ADDR, .size = sizeof(struct in_addr) },
+        [IFA_ANYCAST]           = { .type = NETLINK_TYPE_IN_ADDR, .size = sizeof(struct in_addr) },
```
Comment (yuwata) [inline src/libsystemd/sd-netlink/netlink-types-rtnl.c]: IFA_ANYCAST takes only IPv6. Not IPv4.

File: src/libsystemd/sd-netlink/netlink-types-rtnl.c [hunk 10]
```diff
@@ -902,13 +904,13 @@ DEFINE_TYPE_SYSTEM(rtnl_link);
...
         [IFA_CACHEINFO]         = { .type = NETLINK_TYPE_CACHE_INFO, .size = sizeof(struct ifa_cacheinfo) },
-        [IFA_MULTICAST]         = { .type = NETLINK_TYPE_IN_ADDR },
+        [IFA_MULTICAST]         = { .type = NETLINK_TYPE_IN_ADDR, .size = sizeof(struct in_addr) },
```
Comment (yuwata) [inline src/libsystemd/sd-netlink/netlink-types-rtnl.c]: This is IPv6 only.

File: src/libsystemd/sd-netlink/netlink-types-rtnl.c [hunk 11]
```diff
@@ -940,13 +942,13 @@ static const NLType rtnl_route_metrics_types[] = {
...
-        [RTA_SRC]               = { .type = NETLINK_TYPE_IN_ADDR }, /* 6? */
+        [RTA_DST]               = { .type = NETLINK_TYPE_IN_ADDR, .size = sizeof(struct in_addr) }, /* 6? */
+        [RTA_SRC]               = { .type = NETLINK_TYPE_IN_ADDR, .size = sizeof(struct in_addr) }, /* 6? */
         [RTA_IIF]               = { .type = NETLINK_TYPE_U32 },
         [RTA_OIF]               = { .type = NETLINK_TYPE_U32 },
-        [RTA_GATEWAY]           = { .type = NETLINK_TYPE_IN_ADDR },
+        [RTA_GATEWAY]           = { .type = NETLINK_TYPE_IN_ADDR, .size = sizeof(struct in_addr) },
         [RTA_PRIORITY]          = { .type = NETLINK_TYPE_U32 },
-        [RTA_PREFSRC]           = { .type = NETLINK_TYPE_IN_ADDR }, /* 6? */
+        [RTA_PREFSRC]           = { .type = NETLINK_TYPE_IN_ADDR, .size = sizeof(struct in_addr) }, /* 6? */
```
Comment (yuwata) [inline src/libsystemd/sd-netlink/netlink-types-rtnl.c]: All these take both IPv4 and IPv6. Please drop the change.

File: src/libsystemd/sd-netlink/netlink-types-rtnl.c [hunk 12]
```diff
@@ -971,8 +973,8 @@ static const NLType rtnl_route_types[] = {
 DEFINE_TYPE_SYSTEM(rtnl_route);
 
 static const NLType rtnl_neigh_types[] = {
-        [NDA_DST]               = { .type = NETLINK_TYPE_IN_ADDR },
-        [NDA_LLADDR]            = { .type = NETLINK_TYPE_ETHER_ADDR },
+        [NDA_DST]               = { .type = NETLINK_TYPE_IN_ADDR, .size = sizeof(struct in_addr) },
+        [NDA_LLADDR]            = { .type = NETLINK_TYPE_ETHER_ADDR, .size = ETH_ALEN },
```
Comment (yuwata) [inline src/libsystemd/sd-netlink/netlink-types-rtnl.c]: These take various length. Please drop.

File: src/libsystemd/sd-netlink/netlink-types-rtnl.c [hunk 13]
```diff
@@ -991,8 +993,8 @@ static const NLType rtnl_addrlabel_types[] = {
 DEFINE_TYPE_SYSTEM(rtnl_addrlabel);
 
 static const NLType rtnl_routing_policy_rule_types[] = {
-        [FRA_DST]                 = { .type = NETLINK_TYPE_IN_ADDR },
-        [FRA_SRC]                 = { .type = NETLINK_TYPE_IN_ADDR },
+        [FRA_DST]                 = { .type = NETLINK_TYPE_IN_ADDR, .size = sizeof(struct in_addr) },
+        [FRA_SRC]                 = { .type = NETLINK_TYPE_IN_ADDR, .size = sizeof(struct in_addr) },
```
Comment (yuwata) [inline src/libsystemd/sd-netlink/netlink-types-rtnl.c]: These take both IPv4 and IPv6.

File: src/libsystemd/sd-netlink/netlink-types-rtnl.c [hunk 14]
```diff
@@ -1021,7 +1023,7 @@ static const NLType rtnl_nexthop_types[] = {
         [NHA_GROUP_TYPE]          = { .type = NETLINK_TYPE_U16 },
         [NHA_BLACKHOLE]           = { .type = NETLINK_TYPE_FLAG },
         [NHA_OIF]                 = { .type = NETLINK_TYPE_U32 },
-        [NHA_GATEWAY]             = { .type = NETLINK_TYPE_IN_ADDR },
+        [NHA_GATEWAY]             = { .type = NETLINK_TYPE_IN_ADDR, .size = sizeof(struct in_addr) },
```
Comment (yuwata) [inline src/libsystemd/sd-netlink/netlink-types-rtnl.c]: This takes both IPv4 and IPv6.

**Comment by topimiettinen** [comment]
@yuwata Updated IFA_ANYCAST and IFA_MULTICAST to use in6_addr, otherwise kept only fou, L2TP and bridge changes.

**Comment by topimiettinen** [comment]
ci (opensuse, tumbleweed) fails with `FileNotFoundError: [Errno 2] No such file or directory: '/var/tmp/mkosi-2kbj3psd/root/usr/etc/pam.d/common-auth'`, not related.

File: src/libsystemd/sd-netlink/netlink-types-genl.c (hunk 1 above)
Comment (topimiettinen) [reply src/libsystemd/sd-netlink/netlink-types-genl.c]: removed

File: src/libsystemd/sd-netlink/netlink-types-rtnl.c (hunk 3 above)
Comment (topimiettinen) [reply src/libsystemd/sd-netlink/netlink-types-rtnl.c]: dropped

File: src/libsystemd/sd-netlink/netlink-types-rtnl.c (hunk 5 above)
Comment (topimiettinen) [reply src/libsystemd/sd-netlink/netlink-types-rtnl.c]: dropped

File: src/libsystemd/sd-netlink/netlink-types-rtnl.c (hunk 4 above)
Comment (topimiettinen) [reply src/libsystemd/sd-netlink/netlink-types-rtnl.c]: dropped

File: src/libsystemd/sd-netlink/netlink-types-rtnl.c (hunk 6 above)
Comment (topimiettinen) [reply src/libsystemd/sd-netlink/netlink-types-rtnl.c]: dropped

File: src/libsystemd/sd-netlink/netlink-types-rtnl.c (hunk 7 above)
Comment (topimiettinen) [reply src/libsystemd/sd-netlink/netlink-types-rtnl.c]: dropped

File: src/libsystemd/sd-netlink/netlink-types-rtnl.c (hunk 8 above)
Comment (topimiettinen) [reply src/libsystemd/sd-netlink/netlink-types-rtnl.c]: dropped

File: src/libsystemd/sd-netlink/netlink-types-rtnl.c (hunk 9 above)
Comment (topimiettinen) [reply src/libsystemd/sd-netlink/netlink-types-rtnl.c]: Dropped IFA_BROADCAST, but used  in6_addr for IFA_ANYCAST.

File: src/libsystemd/sd-netlink/netlink-types-rtnl.c (hunk 10 above)
Comment (topimiettinen) [reply src/libsystemd/sd-netlink/netlink-types-rtnl.c]: Changed to in6_addr

File: src/libsystemd/sd-netlink/netlink-types-rtnl.c (hunk 11 above)
Comment (topimiettinen) [reply src/libsystemd/sd-netlink/netlink-types-rtnl.c]: Dropped

File: src/libsystemd/sd-netlink/netlink-types-rtnl.c (hunk 12 above)
Comment (topimiettinen) [reply src/libsystemd/sd-netlink/netlink-types-rtnl.c]: Dropped

File: src/libsystemd/sd-netlink/netlink-types-rtnl.c (hunk 13 above)
Comment (topimiettinen) [reply src/libsystemd/sd-netlink/netlink-types-rtnl.c]: Dropped

File: src/libsystemd/sd-netlink/netlink-types-rtnl.c (hunk 14 above)
Comment (topimiettinen) [reply src/libsystemd/sd-netlink/netlink-types-rtnl.c]: Dropped

File: src/libsystemd/sd-netlink/netlink-types-genl.c (hunk 2 above)
Comment (topimiettinen) [reply src/libsystemd/sd-netlink/netlink-types-genl.c]: Dropped

**Review by yuwata — APPROVED** [review]
LGTM. Thank you.

**Pull request merged**